File: api/src/models/transaction.js
import Sequelize from 'sequelize';

export const definition = [
  'transactions',
  {
    id: { type: Sequelize.INTEGER, autoIncrement: true, primaryKey: true },
    importId: { type: Sequelize.TEXT }, // to differentiate between same transaction imported twice
    receipt: { type: Sequelize.BOOLEAN, defaultValue: false },
    emitterName: { type: Sequelize.TEXT, allowNull: true },
    receiverName: { type: Sequelize.TEXT, allowNull: true },
    description: Sequelize.TEXT,
    comments: Sequelize.TEXT,
    assCard: { type: Sequelize.TEXT, allowNull: true },
    amount: { type: Sequelize.FLOAT, allowNull: false },
    currency: { type: Sequelize.TEXT, allowNull: false },
    date: { type: Sequelize.DATE },
    valueDate: { type: Sequelize.DATE },
    favourite: { type: Sequelize.BOOLEAN, defaultValue: false },
    balance: { type: Sequelize.FLOAT, defaultValue: 0.0 },
    createdAt: { type: Sequelize.DATE, defaultValue: Sequelize.NOW },
    updatedAt: { type: Sequelize.DATE },
    accountId: {
      type: Sequelize.INTEGER,
      allowNull: false,
      references: {
        model: 'accounts',
        key: 'id'
      }
    },
  }, {
    timestamps:true,
  }
];

export default (sequelize) => {
  sequelize.define(...definition);
};

File: backend/src/main/java/com/wheimo/api/domain/entity/Transaction.java
package com.wheimo.api.domain.entity;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "transactions")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class Transaction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "import_id", unique = true)
    private String importId;

    @Column(name = "emitter_name")
    private String emitterName;

    @Column(name = "receiver_name")
    private String receiverName;

    private String description;

    private String comments;

    @Column(name = "ass_card")
    private String assCard;

    @Column(nullable = false)
    private BigDecimal amount;

    @Column(nullable = false, length = 10)
    private String currency;

    private OffsetDateTime date;

    @Column(name = "value_date")
    private OffsetDateTime valueDate;

    @Builder.Default
    private BigDecimal balance = BigDecimal.ZERO;

    @Builder.Default
    private boolean receipt = false;

    @Builder.Default
    private boolean draft = false;

    @Builder.Default
    private boolean favourite = false;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "account_id", nullable = false)
    private Account account;

    @ManyToMany
    @JoinTable(
        name = "transaction_tags",
        joinColumns = @JoinColumn(name = "transaction_id"),
        inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    @Builder.Default
    private List<Tag> tags = new ArrayList<>();

    @OneToMany(mappedBy = "transaction", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<Attachment> attachments = new ArrayList<>();

    @Column(name = "created_at")
    private OffsetDateTime createdAt;

    @Column(name = "updated_at")
    private OffsetDateTime updatedAt;

    @PrePersist
    void prePersist() {
        createdAt = OffsetDateTime.now();
    }

    @PreUpdate
    void preUpdate() {
        updatedAt = OffsetDateTime.now();
    }
}
